nixos/modules/services/hardware/pid-fan-controller.nix

```diff
@@ -0,0 +1,182 @@
+                D = -6.0e-3;
+              };
+            }
+            {
+              name = "gpu";
+              wildcard_path = "/sys/class/drm/card*/device/hwmon/hwmon*/temp2_input";
+              PID_params = {
+                set_point = 65;
+                P = -5.0e-3;
+                I = -2.0e-3;
+                D = -6.0e-3;
+              };
+            }
+          ];
+        '';
+      };
+      fans = lib.mkOption {
+        type = lib.types.listOf (lib.types.submodule fan);
+        description = "list of fans";
+        example = ''
+          [
+            {
+              wildcard_path = "/sys/devices/platform/nct6775.2592/hwmon/hwmon*/pwm1";
+              min_pwm = 60;
+              max_pwm = 255;
+              heat_pressure_srcs = [
+                "cpu"
+                "gpu"
+              ];
+            }
+            {
+              wildcard_path = "/sys/devices/platform/nct6775.2592/hwmon/hwmon*/pwm4";
+              min_pwm = 60;
+              max_pwm = 200;
+              cutoff = true;
+              heat_pressure_srcs = [
+                "cpu"
+              ];
+            }
+          ];
+        '';
+      };
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    environment.etc."pid-fan-settings.json".text = builtins.toJSON cfg.settings;
+
+    systemd.services.pid-fan-controller = {
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = [ "${cfg.package}/bin/pid-fan-controller" ];
+        ExecStopPost = [ "${cfg.package}/bin/pid-fan-controller disable" ];
+        Restart = "always";
+        #This service needs to run as root to write to /sys.
+        #therefore it should operate with the least amount of priviledges needed
+        ProtectHome = "yes";
+        #strict is not possible as it needs /sys
+        ProtectSystem = "full";
+        ProtectProc = "invisible";
+        PrivateNetwork = "yes";
+        NoNewPrivileges = "yes";
+        MemoryDenyWriteExecute = "yes";
+        RestrictNamespaces = "~user pid net uts mnt";
+        ProtectKernelModules = "yes";
+        RestrictRealtime = "yes";
+        SystemCallFilter = "@system-service";
+        CapabilityBoundingSet = "~CAP_KILL CAP_WAKE_ALARM CAP_IPC_LOC CAP_BPF CAP_LINUX_IMMUTABLE CAP_BLOCK_SUSPEND CAP_MKNOD";
+      };
+      # restart unit if config changed
+      restartTriggers = [ config.environment.etc."pid-fan-settings.json".text ];
+    };
+    #sleep hook to restart the service as it breaks sometimes otherwise
+    systemd.services.pid-fan-controller-sleep = {
+      before = [ "sleep.target" ];
+      wantedBy = [ "sleep.target" ];
+      unitConfig = {
+        StopWhenUnneeded = "yes";
+      };
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        ExecStart = [ "${pkgs.systemd}/bin/systemctl stop pid-fan-controller.service" ];
+        ExecStop = [ "${pkgs.systemd}/bin/systemctl restart pid-fan-controller.service" ];
```

Comment:
Isn't systemctl already by default in the PATH? so we could omit the absolute path to make sure we use config.systemd.package.

Reply:
IIRC Exec* only accepts an absolute path, but using `config.systemd.package` is definitely the right way to do this.